cmd/crank/beta/validate/loader.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2024 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validate
+
+import (
+	"bufio"
+	"io"
+	"os"
+	"path/filepath"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/yaml"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+)
+
+// Loader interface defines the contract for different input sources
+type Loader interface {
+	Load() ([]*unstructured.Unstructured, error)
+}
+
+// NewLoader returns a Loader based on the input source
+func NewLoader(input string) (Loader, error) {
+	if input == "-" {
+		return &StdinLoader{}, nil
+	}
+
+	fi, err := os.Stat(input)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot stat input source")
+	}
+
+	if fi.IsDir() {
+		return &FolderLoader{path: input}, nil
+	}
+
+	return &FileLoader{path: input}, nil
+}
+
+// StdinLoader implements the Loader interface for reading from stdin
+type StdinLoader struct{}
+
+// Load reads the contents from stdin
+func (s *StdinLoader) Load() ([]*unstructured.Unstructured, error) {
+	stream, err := load(os.Stdin)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot load YAML stream from stdin")
+	}
+
+	return streamToUnstructured(stream)
+}
+
+// FileLoader implements the Loader interface for reading from a file and converting input to unstructured objects
+type FileLoader struct {
+	path string
+}
+
+// Load reads the contents from a file
+func (f *FileLoader) Load() ([]*unstructured.Unstructured, error) {
+	stream, err := readFile(f.path)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot read file")
+	}
+
+	return streamToUnstructured(stream)
+}
+
+// FolderLoader implements the Loader interface for reading from a folder
+type FolderLoader struct {
+	path string
+}
+
+// Load reads the contents from all files in a folder
+func (f *FolderLoader) Load() ([]*unstructured.Unstructured, error) {
+	var stream [][]byte
+	err := filepath.Walk(f.path, func(path string, info os.FileInfo, err error) error {
```

Comment:
Could make sense to only read `.yaml` or `.yml` files, we usually do that and it's a pity one has to change the repository structure just because for example there is a `.md` file in the same folder where resources or extensions live.

Reply:
e.g. <img width="904" alt="image" src="https://github.com/crossplane/crossplane/assets/5697904/5616bd71-5d02-46c2-91c2-e78b0814604c">

Reply:
added file extension check